File: Blind75/39. Combination Sum/39. Combination Sum.cpp
class Solution {
public:
    vector<vector<int>> res;
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> tmp;
        backtracking(candidates, tmp, 0, 0, target);
        return res;
    }

    void backtracking(vector<int>& candidates, vector<int>& tmp, int index, int sum, int target){
        //base case
        int n = candidates.size();
        if(target==sum){
            res.push_back(tmp);
            return;
        }
        if(target<sum) return;

        for(int i=index; i<n; i++){
            tmp.push_back(candidates[i]);
            sum+=candidates[i];
            backtracking(candidates,tmp,i,sum,target);
            sum-=candidates[i];
            tmp.pop_back();
        }
    }
};

File: Blind75/39. Combination Sum/39. Combination Sum.cs
public class Solution {
    IList<IList<int>> res = new List<IList<int>>();
    public IList<IList<int>> CombinationSum(int[] candidates, int target) {
        //backtracking
        List<int> tmp = new List<int>();
        backtracking(candidates, target, ref tmp, 0, 0);
        return res;
    }

    void backtracking(int[] candidates, int target, ref List<int> tmp, int index, int sum){
        //basecase
        if(sum==target){
            res.Add(tmp.ToList());
            return;
        }
        if(sum>target) return;

        for(int i=index; i<candidates.Length; i++){
            tmp.Add(candidates[i]);
            sum+=candidates[i];
            backtracking(candidates, target, ref tmp, i, sum);
            sum-=candidates[i];
            tmp.RemoveAt(tmp.Count-1);
        }
    }
}
